File: backend/api/v1/schemas/playlists.py
from typing import Annotated, Literal

import msgspec
from infrastructure.msgspec_fastapi import AppStruct


class PlaylistTrackResponse(AppStruct):
    id: str
    position: int
    track_name: str
    artist_name: str
    album_name: str
    album_id: str | None = None
    artist_id: str | None = None
    track_source_id: str | None = None
    cover_url: str | None = None
    source_type: str = ""
    available_sources: list[str] | None = None
    format: str | None = None
    track_number: int | None = None
    disc_number: int | None = None
    duration: int | None = None
    created_at: str = ""
    plex_rating_key: str | None = None


class PlaylistSummaryResponse(AppStruct):
    id: str
    name: str
    track_count: int = 0
    total_duration: int | None = None
    cover_urls: list[str] = msgspec.field(default_factory=list)
    custom_cover_url: str | None = None
    source_ref: str | None = None
    created_at: str = ""
    updated_at: str = ""
    # Ownership / visibility (D4).
    is_public: bool = False
    is_owner: bool = False
    owner_name: str | None = None
    is_redacted: bool = False


class PlaylistDetailResponse(AppStruct):
    # Keep these fields in sync with PlaylistSummaryResponse because the frontend extends PlaylistSummary.
    id: str
    name: str
    cover_urls: list[str] = msgspec.field(default_factory=list)
    custom_cover_url: str | None = None
    source_ref: str | None = None
    tracks: list[PlaylistTrackResponse] = msgspec.field(default_factory=list)
    track_count: int = 0
    total_duration: int | None = None
    created_at: str = ""
    updated_at: str = ""
    # Ownership / visibility (D4).
    is_public: bool = False
    is_owner: bool = False
    owner_name: str | None = None
    is_redacted: bool = False


class RedactedPlaylist(AppStruct):
    """Admin's view of another user's PRIVATE playlist (D4): existence + count + owner
    only, never the name/tracks/covers. Returned in the list and as a detail body (200)."""
    id: str
    track_count: int = 0
    owner_name: str | None = None
    is_redacted: bool = True


class PlaylistListResponse(AppStruct):
    playlists: list[PlaylistSummaryResponse | RedactedPlaylist] = msgspec.field(default_factory=list)


class SetPlaylistPublicRequest(AppStruct):
    is_public: bool


class CreatePlaylistRequest(AppStruct):
    name: str


class UpdatePlaylistRequest(AppStruct):
    name: str | None = None


class SmartMixSeed(AppStruct):
    type: Literal["artist", "genre", "mood"]
    value: str


class GeneratePlaylistRequest(AppStruct):
    """Smart Mix: generate and persist a playlist from a blend of seeds
    (1-10, any mix of artist/genre/mood)."""
    seeds: Annotated[list[SmartMixSeed], msgspec.Meta(min_length=1, max_length=10)]
    count: Annotated[int, msgspec.Meta(ge=1, le=250)] = 25
    name: str | None = None


class TrackDataRequest(AppStruct):
    track_name: str
    artist_name: str
    album_name: str
    album_id: str | None = None
    artist_id: str | None = None
    track_source_id: str | None = None
    cover_url: str | None = None
    source_type: str = ""
    available_sources: list[str] | None = None
    format: str | None = None
    track_number: int | None = None
    disc_number: int | None = None
    duration: float | int | None = None
    plex_rating_key: str | None = None


class AddTracksRequest(AppStruct):
    tracks: list[TrackDataRequest]
    position: int | None = None


class RemoveTracksRequest(AppStruct):
    track_ids: list[str]


class ReorderTrackRequest(AppStruct):
    track_id: str
    new_position: int


class ReorderTrackResponse(AppStruct):
    status: str = "ok"
    message: str = "Track reordered"
    actual_position: int = 0


class UpdateTrackRequest(AppStruct):
    source_type: str | None = None
    available_sources: list[str] | None = None


class AddTracksResponse(AppStruct):
    tracks: list[PlaylistTrackResponse] = msgspec.field(default_factory=list)


class CoverUploadResponse(AppStruct):
    cover_url: str


class TrackIdentifier(AppStruct):
    track_name: str
    artist_name: str
    album_name: str


class CheckTrackMembershipRequest(AppStruct):
    tracks: list[TrackIdentifier]


class CheckTrackMembershipResponse(AppStruct):
    membership: dict[str, list[int]]


class ResolveSourcesResponse(AppStruct):
    sources: dict[str, list[str]]

File: frontend/src/lib/api/playlists.ts
import { API } from '$lib/constants';
import { api } from '$lib/api/client';
import type { QueueItem } from '$lib/player/types';

export interface PlaylistTrack {
	id: string;
	position: number;
	track_name: string;
	artist_name: string;
	album_name: string;
	album_id: string | null;
	artist_id: string | null;
	track_source_id: string | null;
	cover_url: string | null;
	source_type: string;
	available_sources: string[] | null;
	format: string | null;
	track_number: number | null;
	disc_number: number | null;
	duration: number | null;
	created_at: string;
	plex_rating_key: string | null;
}

export interface PlaylistSummary {
	id: string;
	name: string;
	track_count: number;
	total_duration: number | null;
	cover_urls: string[];
	custom_cover_url: string | null;
	source_ref: string | null;
	created_at: string;
	updated_at: string;
	// Ownership / visibility (D4).
	is_public: boolean;
	is_owner: boolean;
	owner_name: string | null;
	is_redacted: false;
}

export interface PlaylistDetail extends PlaylistSummary {
	tracks: PlaylistTrack[];
}

/** Admin's view of another user's PRIVATE playlist (D4): count + owner only. */
export interface RedactedPlaylist {
	id: string;
	track_count: number;
	owner_name: string | null;
	is_redacted: true;
}

export type PlaylistListItem = PlaylistSummary | RedactedPlaylist;
export type PlaylistDetailItem = PlaylistDetail | RedactedPlaylist;

export function isRedactedPlaylist(
	item: PlaylistListItem | PlaylistDetailItem
): item is RedactedPlaylist {
	return item.is_redacted === true;
}

export interface TrackData {
	track_name: string;
	artist_name: string;
	album_name: string;
	album_id?: string | null;
	artist_id?: string | null;
	track_source_id?: string | null;
	cover_url?: string | null;
	source_type: string;
	available_sources?: string[] | null;
	format?: string | null;
	track_number?: number | null;
	disc_number?: number | null;
	duration?: number | null;
	plex_rating_key?: string | null;
}

export function queueItemToTrackData(item: QueueItem): TrackData {
	return {
		track_name: item.trackName,
		artist_name: item.artistName,
		album_name: item.albumName,
		album_id: item.albumId || null,
		artist_id: item.artistId || null,
		track_source_id: item.trackSourceId || null,
		cover_url: item.coverUrl,
		source_type: item.sourceType,
		available_sources: item.availableSources ?? null,
		format: item.format ?? null,
		track_number: item.trackNumber ?? null,
		disc_number: item.discNumber ?? null,
		duration: item.duration ?? null,
		plex_rating_key: item.plexRatingKey ?? null
	};
}

export async function fetchPlaylists(): Promise<PlaylistListItem[]> {
	const data = await api.global.get<{ playlists: PlaylistListItem[] }>(API.playlists.list());
	return data.playlists;
}

export async function fetchPlaylist(
	id: string,
	options?: { signal?: AbortSignal }
): Promise<PlaylistDetailItem> {
	return api.global.get<PlaylistDetailItem>(API.playlists.detail(id), { signal: options?.signal });
}

export async function setPlaylistPublic(id: string, isPublic: boolean): Promise<PlaylistSummary> {
	return api.global.patch<PlaylistSummary>(API.playlists.share(id), { is_public: isPublic });
}

export async function createPlaylist(name: string): Promise<PlaylistDetail> {
	return api.global.post<PlaylistDetail>(API.playlists.create(), { name });
}

export type SmartMixSeedType = 'artist' | 'genre' | 'mood';

export interface SmartMixSeed {
	type: SmartMixSeedType;
	value: string;
}

export interface GenerateSmartMixInput {
	seeds: SmartMixSeed[];
	count?: number;
	name?: string;
}

/** Smart Mix: generate and persist a playlist from a blend of artist / genre / mood seeds. */
export async function generateSmartMix(input: GenerateSmartMixInput): Promise<PlaylistDetail> {
	return api.global.post<PlaylistDetail>(API.playlists.generate(), input);
}

export async function updatePlaylist(id: string, data: { name?: string }): Promise<PlaylistDetail> {
	return api.global.put<PlaylistDetail>(API.playlists.update(id), data);
}

export async function deletePlaylist(id: string): Promise<void> {
	await api.global.delete(API.playlists.delete(id));
}

export async function addTracksToPlaylist(
	id: string,
	tracks: TrackData[],
	position?: number
): Promise<PlaylistTrack[]> {
	const body: { tracks: TrackData[]; position?: number } = { tracks };
	if (position != null) body.position = position;
	const data = await api.global.post<{ tracks: PlaylistTrack[] }>(
		API.playlists.addTracks(id),
		body
	);
	return data.tracks;
}

export async function removeTrackFromPlaylist(id: string, trackId: string): Promise<void> {
	await api.global.delete(API.playlists.removeTrack(id, trackId));
}

export async function removeTracksFromPlaylist(id: string, trackIds: string[]): Promise<void> {
	await api.global.post(API.playlists.removeTracks(id), { track_ids: trackIds });
}

export async function updatePlaylistTrack(
	id: string,
	trackId: string,
	data: { source_type?: string; available_sources?: string[] }
): Promise<PlaylistTrack> {
	return api.global.patch<PlaylistTrack>(API.playlists.updateTrack(id, trackId), data);
}

export async function reorderPlaylistTrack(
	id: string,
	trackId: string,
	newPosition: number
): Promise<{ actual_position: number }> {
	return api.global.patch<{ actual_position: number }>(API.playlists.reorderTrack(id), {
		track_id: trackId,
		new_position: newPosition
	});
}

export async function uploadPlaylistCover(id: string, file: File): Promise<{ cover_url: string }> {
	const formData = new FormData();
	formData.append('cover_image', file);
	return api.global.upload<{ cover_url: string }>(API.playlists.uploadCover(id), formData);
}

export async function deletePlaylistCover(id: string): Promise<void> {
	await api.global.delete(API.playlists.deleteCover(id));
}

export async function checkTrackMembership(
	tracks: { track_name: string; artist_name: string; album_name: string }[]
): Promise<Record<string, number[]>> {
	const data = await api.global.post<{ membership: Record<string, number[]> }>(
		API.playlists.checkTracks(),
		{ tracks }
	);
	return data.membership;
}

export async function resolvePlaylistSources(id: string): Promise<Record<string, string[]>> {
	const data = await api.global.post<{ sources: Record<string, string[]> }>(
		API.playlists.resolveSources(id)
	);
	return data.sources;
}

export interface BatchRequestResult {
	success: boolean;
	message: string;
	requested: number;
	skipped: number;
}

export async function requestMissingTracks(id: string): Promise<BatchRequestResult> {
	return api.global.post<BatchRequestResult>(API.playlists.requestMissing(id));
}
